man/bzadmin.6.in
.\" bzflag
.\" Copyright (c) 1993-2025 Tim Riker
.\"
.\" This package is free software;  you can redistribute it and/or
.\" modify it under the terms of the license found in the file
.\" named COPYING that should have accompanied this file.
.\"
.\" THIS PACKAGE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.TH "BZADMIN" "6" "@BUILD_DATE@" "@PACKAGE_STRING@" "BZFlag"
.SH "NAME"
bzadmin \- a text based client for BZFlag
.SH "SYNOPSIS"
.B bzadmin
[\fB\-help\fR]
[\fB\-hide \fR\fImsgtype\fR[\fI,msgtype\fR]...\fR]
[\fB\-show \fR\fImsgtype\fR[\fI,msgtype\fR]...\fR]
[\fB\-ui \fR{\fIcurses\fR | \fIstdboth\fR | \fIstdin\fR | \fIstdout\fR}]
\fIcallsign\fR[\fI:password\fR]\fI@hostname\fR[\fI:port\fR] [\fIcommand\fR] [\fIcommand\fR] ...
.SH "DESCRIPTION"
.B bzadmin
is a textbased client for the game BZFlag. It can't be used for
playing, but it can be used to see when players join and leave the
game, to see the chat messages, and to send messages and commands
to the server, as well as see several other vital game messages.
.PP
When you start bzadmin without any command line options other than
callsign and hostname a simple curses\-based user interface will be
started (unless you built bzadmin without curses support). This
interface is divided into three rectangles; the output window
(which covers almost all of the terminal), the target window, and
the input window.
.PP
The output window is where messages from the server will be shown.
.PP
The target window shows the name of the player that
will receive your next message, or 'all' if it will be a public message.
You can change target by using the left and right arrow keys.
.PP
The input window is where you type your messages. It also supports tab
completion of commands and callsigns. You can clear the input window
with the key combination \fBCtrl\-U\fR, and you can generate a \fB/kick\fR
command for the current target with the F5 key (if the current target
is a player). You can also generate a ban command for the current target with
the F6 key, but this only works if you are an admin and the server has sent
you that players IP address (as a response to a \fB/playerlist\fR command).
.PP
The curses user interface also has a simple menu system where you can edit
the server variables (if you are an admin) and see a player list. If you are
an admin you can also see the IP addresses in the player list, if you have
sent a \fB/playerlist\fR command. The menu shows up when you hit the F2 key.
It only covers the upper half of the screen, so you can still see what's
going on in the game. You can navigate through the menu with the up and
down arrow keys, and use the enter key to select submenus and other menu
items. If you hit F2 again the command prompt will regain keyboard focus,
but the menu will still be visible. If you hit F2 a third time the menu will
go away.
.SS Options
.RS
.TP 15
.B \-help
Show a simple help text.
.TP
\fB\-hide \fR\fImsgtype\fR[\fI,msgtype\fR]...
Tell bzadmin not to show these message types. The available message types are
chat, join, kill, leave, pause, ping, rabbit, spawn, time, and over. By default
chat, join, kill, leave, pause, and rabbit are shown. You can also change this
while bzadmin is running with the commands \fB/show <msgtype>\fR and
\fB/hide <msgtype>\fR, and from the curses menu.
.TP
\fB\-show \fR\fImsgtype\fR[\fI,msgtype\fR]...
Tell bzadmin to show these message types. See \fB\-hide\fR for a list of
available message types. If a message type is listed both in \fB\-show\fR
and \fB\-hide\fR it will not be shown.
.TP
\fB\-ui \fR{\fIcurses\fR | \fIstdboth\fR | \fIstdin\fR | \fIstdout\fR}
Select the user interface that you want. The curses interface is the default,
and it is described above.
.br
The stdin interface reads user commands from the standard in stream
and sends them to the server. All server output is ignored.
.br
The stdout interface prints all server output to the standard out stream.
All user input is ignored.
.br
The stdboth interface is a combination of stdin and stdout \- it prints
server output to the standard out stream, and reads user commands from
the standard in stream.
.TP
\fIcallsign\fR[\fI:password\fR]\fI@hostname\fR[\fI:port\fR]
Specifies the callsign that you want your client to use, and the
hostname where the BZFlag server is. The port number is optional,
and the default value is 5154.
.TP
\fIcommand\fR
You can specify messages and commands to send to the server on the
command line. BZAdmin will continue running after sending the
specified commands.
.RE
.SS Examples
.RS
.TP 15
\fBbzadmin admin@localhost:5154\fR
.br
Join the game on localhost, port 5154, using the callsign 'admin'.
.TP
\fBbzadmin admin@localhost '/passwd secretpass' '/ban 192.168.0.2'
.br
Connect to the server at localhost and ban the IP 192.168.0.2.
.TP
\fBbzadmin \-ui stdout spy@bzserver.xy | grep magicword\fR
Connect to bzserver.xy and print all server messages that contain 'magicword'.
.RE
.SH "SEE ALSO"
bzfs(6), bzflag(6), bzw(5)


NEWS
see version entry in ChangeLog



src/bzadmin/bzadmin.cxx
/* bzflag
 * Copyright (c) 1993-2025 Tim Riker
 *
 * This package is free software;  you can redistribute it and/or
 * modify it under the terms of the license found in the file
 * named COPYING that should have accompanied this file.
 *
 * THIS PACKAGE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

#include "common.h"

/* system headers */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include <map>
#include <string>

/* common headers */
#include "TimeKeeper.h"

/* local headers */
#include "BZAdminClient.h"
#include "BZAdminUI.h"
#include "OptionParser.h"
#include "UIMap.h"


// causes persistent rebuilding to obtain build versioning
#include "version.h"

int debugLevel = 0;

/** @file
    This is the main file for bzadmin, the bzflag text client.
*/


int main(int argc, char** argv)
{

#ifdef _WIN32
    // startup winsock
    {
        static const int major = 2, minor = 2;
        WSADATA wsaData;
        if (WSAStartup(MAKEWORD(major, minor), &wsaData))
        {
            std::cerr << "Could not initialise WinSock.";
            return 1;
        }
        if (LOBYTE(wsaData.wVersion) != major || HIBYTE(wsaData.wVersion) != minor)
        {
            std::cerr << "Invalid WinSock version (got " << (int) LOBYTE(wsaData.wVersion) <<
                      '.' << (int) HIBYTE(wsaData.wVersion) << ", expected" << major << '.' << minor << ')';
            WSACleanup();
            return 1;
        }
    }
#endif
    // command line options
    std::string uiName("curses");
    std::vector<std::string> visibleMsgs;
    std::vector<std::string> invisibleMsgs;

    // no curses, use stdboth as default instead
    const UIMap& interfaces = UIMap::instance();
    if (interfaces.find("curses") == interfaces.end())
        uiName = "stdboth";

    // build a usage string with all interfaces
    UIMap::const_iterator uiIter;
    std::string uiUsage;
    for (uiIter = interfaces.begin(); uiIter != interfaces.end(); ++uiIter)
        uiUsage += uiIter->first + '|';
    uiUsage = std::string("[-ui ") + uiUsage.substr(0, uiUsage.size() - 1) + ']';

    // register and parse command line arguments
    OptionParser op(std::string("bzadmin ") + getAppVersion(),
                    "CALLSIGN[:password]@HOST[:PORT] [COMMAND] [COMMAND] ...");
    const std::string uiOption("ui");
    const std::string uiMsg = "choose a user interface";
    op.registerVariable(uiOption, uiName, uiUsage, uiMsg);
    op.registerVariable("list", startupInfo.listServerURL, "[-list <list-server-url>]", "specify a list server to use");
    op.registerVector("show", visibleMsgs, "[-show msgtype{,msgtype}*]",
                      "tell bzadmin to show these message types");
    op.registerVector("hide", invisibleMsgs, "[-hide msgtype{,msgtype}*]",
                      "tell bzadmin not to show these message types");
    if (!op.parse(argc, argv))
        return 1;

    // check that the ui is valid
    uiIter = UIMap::instance().find(uiName);
    if (uiIter == UIMap::instance().end())
    {
        std::cerr<<"There is no interface called \""<<uiName<<"\"."<<std::endl;
        return 1;
    }

    // check that we have callsign and host in the right format and extract them
    {
        int atPos;
        std::string callsign = "", password = "", serverName = "";
        if (!(op.getParameters().size() > 0 &&
                (atPos = op.getParameters()[0].find('@')) > 0))
        {
            // input callsign and host interactively
            std::cout << "No callsign@host specified.  Please input them" << std::endl;
            std::cout << "Callsign: ";
            std::getline(std::cin, callsign);
            if (callsign.size() <= 1)
            {
                std::cerr << "You must specify a callsign.  Exiting." << std::endl;
                return 1;
            }
            std::cout << "Password (optional): ";
            std::getline(std::cin, password);
            if (password.size() <= 1)
                std::cerr << "Not using central login" << std::endl;
            std::cout << "Server[:port] to connect to: ";
            std::getline(std::cin, serverName);
            if (serverName.size() <= 1)
            {
                std::cerr << "You must specify a host name to connect to.  Exiting." << std::endl;
                return 1;
            }
        }
        else     // callsign:password@host:port on command line
        {
            callsign = op.getParameters()[0].substr(0, atPos);
            int pPos = callsign.find(':');
            if (pPos != -1)
            {
                password = callsign.substr(pPos + 1).c_str();
                callsign = callsign.substr(0, pPos);
            }
            serverName = op.getParameters()[0].substr(atPos + 1);
        }
        startupInfo.serverPort = ServerPort;
        int cPos = serverName.find(':');
        if (cPos != -1)
        {
            long int serverPort = strtol(serverName.substr(cPos + 1).c_str(), (char **)NULL, 10);
            if (serverPort > 0 && serverPort < 65536)
                startupInfo.serverPort = (int) serverPort;
            serverName = serverName.substr(0, cPos);
        }
        // Flawfinder: ignore
        strncpy(startupInfo.callsign, callsign.c_str(), sizeof(startupInfo.callsign) - 1);
        // Flawfinder: ignore
        strncpy(startupInfo.password, password.c_str(), sizeof(startupInfo.password) - 1);
        // Flawfinder: ignore
        strncpy(startupInfo.serverName, serverName.c_str(), sizeof(startupInfo.serverName) - 1);
    }
    std::cerr << "Connecting to " <<
              startupInfo.callsign << "@" <<
              startupInfo.serverName << ":" <<
              startupInfo.serverPort;
    // Check if password is not empty
    if (startupInfo.password[0])
        std::cerr << " using central login";
    std::cerr << std::endl;

    // try to connect
    BZAdminClient client;
    if (!client.isValid())
        return 1;

    unsigned int i;
    for (i = 0; i < visibleMsgs.size(); ++i)
        client.showMessageType(visibleMsgs[i]);
    for (i = 0; i < invisibleMsgs.size(); ++i)
        client.ignoreMessageType(invisibleMsgs[i]);

    // if we got commands as arguments, send them
    if (op.getParameters().size() > 1)
    {
        // if we have a token wait a bit for global login
        // FIXME: should "know" when we are logged in (or fail) and only wait that long.
        if (startupInfo.token[0] != 0)
            TimeKeeper::sleep(5.0);
        for (unsigned int j = 1; j < op.getParameters().size(); ++j)
        {
            const std::string& cmd = op.getParameters()[j];
            if (cmd == "/quit")
            {
                client.waitForServer();
                return 0;
            }
            else if (strncasecmp(cmd.c_str(), "/sleep", 6) == 0)
            {
                const char* start = cmd.c_str() + 6;
                char* endptr;
                double sleepTime = strtod(start, &endptr);
                if (endptr != start)
                    TimeKeeper::sleep(sleepTime);
            }
            else
                client.sendMessage(cmd, AllPlayers);
        }
    }

    // create UI and run the main loop
    BZAdminUI* ui = uiIter->second(client);
    client.setUI(ui);
    client.runLoop();
    delete ui;

    return 0;
}

// Local Variables: ***
// mode: C++ ***
// tab-width: 4 ***
// c-basic-offset: 4 ***
// indent-tabs-mode: nil ***
// End: ***
// ex: shiftwidth=4 tabstop=4
